lib/connectors/base/simple_rules_parser.rb

```diff
@@ -11,23 +11,144 @@
 
 module Connectors
   module Base
+    class FilteringRulesValidationError < StandardError; end
+
     class SimpleRulesParser
+
+      include Core::Filtering
+
+      attr_reader :rules
+
       def initialize(rules)
-        @rules = (rules || []).map(&:with_indifferent_access).filter { |r| r[:id] != 'DEFAULT' }.sort_by { |r| r[:order] }
+        begin
+          sorted = (rules || []).map { |r| SimpleRule.new(r) }.filter { |r| r.id != 'DEFAULT' }.sort_by(&:order)
+        rescue StandardError => e
+          raise FilteringRulesValidationError, "Invalid rule: #{e.message}"
+        end
+        @rules = validate(sorted)
       end
 
       def parse
-        merge_rules(@rules.map do |rule_hash|
-          rule = Core::Filtering::SimpleRule.new(rule_hash)
-          unless rule.is_include? || rule.is_exclude?
-            raise "Unknown policy: #{rule.policy}"
-          end
+        merge_rules(@rules.map do |rule|
           parse_rule(rule)
         end)
       end
 
+      def validate(rules)
+        return rules if rules.empty?
+        rules.each do |rule|
+          validate_rule(rule)
+        end
+        field_acc = {}
+        rules.each do |rule|
+          if field_acc[rule.field].present?
+            field_acc[rule.field] << rule
+          else
+            field_acc[rule.field] = [rule]
+          end
+        end
+        result = []
+        field_acc.each_value do |field_rules|
+          result << filter_field_rules(field_rules)
+        end
+        result.flatten
+      end
+
       private
 
+      def filter_field_rules(field_rules)
+        if field_rules.size <= 1
+          return field_rules
+        end
+        # drop contradicting equality rules
+        result = drop_invalid_equality_rules(field_rules)
+        if result.size > 1
+          result = drop_invalid_starts_with_rules(result)
+        end
+        if result.size > 1
+          result = drop_invalid_ends_with_rules(result)
+        end
+
+        # # check for overlapping ranges
+        # ranges = field_rules.filter { |r| r[:rule] == '>' || r[:rule] == '<' }
+        # ranges.each_with_index do |r, i|
+        #   next if i == ranges.size - 1
+        #   next_r = ranges[i + 1]
+        #   if r[:value] == next_r[:value]
+        #     raise FilteringRulesValidationError.new("Contradicting rules for field: #{field}. Can't have overlapping ranges.")
+        #   end
```

Comment:
should this block be removed?

Reply:
Yeah it should, it's a WIP to document the thought process. Frankly, the ranges feel so cumbersome to validate... there's just too many cases.